# Исследование надежности заемщиков


Во второй части проекта вы выполните шаги 3 и 4. Их вручную проверит ревьюер.
Чтобы вам не пришлось писать код заново для шагов 1 и 2, мы добавили авторские решения в ячейки с кодом. 



## Откройте таблицу и изучите общую информацию о данных

**Задание 1. Импортируйте библиотеку pandas. Считайте данные из csv-файла в датафрейм и сохраните в переменную `data`. Путь к файлу:**

`/datasets/data.csv`

In [36]:
import pandas as pd

try:
    data = pd.read_csv('/datasets/data.csv')
except:
    data = pd.read_csv('https://code.s3.yandex.net/datasets/data.csv')

**Задание 2. Выведите первые 20 строчек датафрейма `data` на экран.**

In [37]:
data.head(20)

children  days_employed  dob_years            education  education_id  \
0          1   -8437.673028         42               высшее             0   
1          1   -4024.803754         36              среднее             1   
2          0   -5623.422610         33              Среднее             1   
3          3   -4124.747207         32              среднее             1   
4          0  340266.072047         53              среднее             1   
5          0    -926.185831         27               высшее             0   
6          0   -2879.202052         43               высшее             0   
7          0    -152.779569         50              СРЕДНЕЕ             1   
8          2   -6929.865299         35               ВЫСШЕЕ             0   
9          0   -2188.756445         41              среднее             1   
10         2   -4171.483647         36               высшее             0   
11         0    -792.701887         40              среднее             1   
12         0            NaN         65              среднее             1   
13         0   -1846.641941         54  неоконченное высшее             2   
14         0   -1844.956182         56               высшее             0   
15         1    -972.364419         26              среднее             1   
16         0   -1719.934226         35              среднее             1   
17         0   -2369.999720         33               высшее             0   
18         0  400281.136913         53              среднее             1   
19         0  -10038.818549         48              СРЕДНЕЕ             1   

       family_status  family_status_id gender income_type  debt  \
0    женат / замужем                 0      F   сотрудник     0   
1    женат / замужем                 0      F   сотрудник     0   
2    женат / замужем                 0      M   сотрудник     0   
3    женат / замужем                 0      M   сотрудник     0   
4   гражданский брак                 1      F   пенсионер     0   
5   гражданский брак                 1      M   компаньон     0   
6    женат / замужем                 0      F   компаньон     0   
7    женат / замужем                 0      M   сотрудник     0   
8   гражданский брак                 1      F   сотрудник     0   
9    женат / замужем                 0      M   сотрудник     0   
10   женат / замужем                 0      M   компаньон     0   
11   женат / замужем                 0      F   сотрудник     0   
12  гражданский брак                 1      M   пенсионер     0   
13   женат / замужем                 0      F   сотрудник     0   
14  гражданский брак                 1      F   компаньон     1   
15   женат / замужем                 0      F   сотрудник     0   
16   женат / замужем                 0      F   сотрудник     0   
17  гражданский брак                 1      M   сотрудник     0   
18    вдовец / вдова                 2      F   пенсионер     0   
19         в разводе                 3      F   сотрудник     0   

     total_income                                 purpose  
0   253875.639453                           покупка жилья  
1   112080.014102                 приобретение автомобиля  
2   145885.952297                           покупка жилья  
3   267628.550329              дополнительное образование  
4   158616.077870                         сыграть свадьбу  
5   255763.565419                           покупка жилья  
6   240525.971920                       операции с жильем  
7   135823.934197                             образование  
8    95856.832424                   на проведение свадьбы  
9   144425.938277                 покупка жилья для семьи  
10  113943.491460                    покупка недвижимости  
11   77069.234271       покупка коммерческой недвижимости  
12            NaN                         сыграть свадьбу  
13  130458.228857                 приобретение автомобиля  
14  165127.911772              покупка жилой недвижимости  
15  116820.904450 

**Задание 3. Выведите основную информацию о датафрейме с помощью метода `info()`.**

In [38]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


## Предобработка данных

### Удаление пропусков

**Задание 4. Выведите количество пропущенных значений для каждого столбца. Используйте комбинацию двух методов.**

In [39]:
data.isna().sum()

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

**Задание 5. В двух столбцах есть пропущенные значения. Один из них — `days_employed`. Пропуски в этом столбце вы обработаете на следующем этапе. Другой столбец с пропущенными значениями — `total_income` — хранит данные о доходах. На сумму дохода сильнее всего влияет тип занятости, поэтому заполнить пропуски в этом столбце нужно медианным значением по каждому типу из столбца `income_type`. Например, у человека с типом занятости `сотрудник` пропуск в столбце `total_income` должен быть заполнен медианным доходом среди всех записей с тем же типом.**

In [10]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['total_income'].isna()), 'total_income'] = \
    data.loc[(data['income_type'] == t), 'total_income'].median()

### Обработка аномальных значений

**Задание 6. В данных могут встречаться артефакты (аномалии) — значения, которые не отражают действительность и появились по какой-то ошибке. таким артефактом будет отрицательное количество дней трудового стажа в столбце `days_employed`. Для реальных данных это нормально. Обработайте значения в этом столбце: замените все отрицательные значения положительными с помощью метода `abs()`.**

In [11]:
data['days_employed'] = data['days_employed'].abs()

**Задание 7. Для каждого типа занятости выведите медианное значение трудового стажа `days_employed` в днях.**

In [12]:
data.groupby('income_type')['days_employed'].agg('median')

income_type
безработный        366413.652744
в декрете            3296.759962
госслужащий          2689.368353
компаньон            1547.382223
пенсионер          365213.306266
предприниматель       520.848083
сотрудник            1574.202821
студент               578.751554
Name: days_employed, dtype: float64

У двух типов (безработные и пенсионеры) получатся аномально большие значения. Исправить такие значения сложно, поэтому оставьте их как есть. Тем более этот столбец не понадобится вам для исследования.

**Задание 8. Выведите перечень уникальных значений столбца `children`.**

In [13]:
data['children'].unique()

array([ 1,  0,  3,  2, -1,  4, 20,  5])

**Задание 9. В столбце `children` есть два аномальных значения. Удалите строки, в которых встречаются такие аномальные значения из датафрейма `data`.**

In [14]:
data = data[(data['children'] != -1) & (data['children'] != 20)]

**Задание 10. Ещё раз выведите перечень уникальных значений столбца `children`, чтобы убедиться, что артефакты удалены.**

In [15]:
data['children'].unique()

array([1, 0, 3, 2, 4, 5])

### Удаление пропусков (продолжение)

**Задание 11. Заполните пропуски в столбце `days_employed` медианными значениями по каждого типа занятости `income_type`.**

In [16]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['days_employed'].isna()), 'days_employed'] = \
    data.loc[(data['income_type'] == t), 'days_employed'].median()

**Задание 12. Убедитесь, что все пропуски заполнены. Проверьте себя и ещё раз выведите количество пропущенных значений для каждого столбца с помощью двух методов.**

In [17]:
data.isna().sum()

children            0
days_employed       0
dob_years           0
education           0
education_id        0
family_status       0
family_status_id    0
gender              0
income_type         0
debt                0
total_income        0
purpose             0
dtype: int64

### Изменение типов данных

**Задание 13. Замените вещественный тип данных в столбце `total_income` на целочисленный с помощью метода `astype()`.**

In [18]:
data['total_income'] = data['total_income'].astype(int)

### Обработка дубликатов

**Задание 14. Обработайте неявные дубликаты в столбце `education`. В этом столбце есть одни и те же значения, но записанные по-разному: с использованием заглавных и строчных букв. Приведите их к нижнему регистру. Проверьте остальные столбцы.**

In [19]:
data['education'] = data['education'].str.lower()

**Задание 15. Выведите на экран количество строк-дубликатов в данных. Если такие строки присутствуют, удалите их.**

In [20]:
data.duplicated().sum()

71

In [21]:
data = data.drop_duplicates()

### Категоризация данных

**Задание 16. На основании диапазонов, указанных ниже, создайте в датафрейме `data` столбец `total_income_category` с категориями:**

- 0–30000 — `'E'`;
- 30001–50000 — `'D'`;
- 50001–200000 — `'C'`;
- 200001–1000000 — `'B'`;
- 1000001 и выше — `'A'`.


**Например, кредитополучателю с доходом 25000 нужно назначить категорию `'E'`, а клиенту, получающему 235000, — `'B'`. Используйте собственную функцию с именем `categorize_income()` и метод `apply()`.**

In [22]:
def categorize_income(income):
    try:
        if 0 <= income <= 30000:  
            return 'E: меньше 30 000' # добавила текстовые описания категорий для упрощения визуализации при анализе категорий доходов
        elif 30001 <= income <= 50000:
            return 'D: от 30 000 до 50 000'
        elif 50001 <= income <= 200000:
            return 'C:от 50 000 до 200 000'
        elif 200001 <= income <= 1000000:
            return 'B от 200 000 до 1 000 000'
        elif income >= 1000001:
            return 'A: больше 1 000 000'
    except:
        pass

In [23]:
data['total_income_category'] = data['total_income'].apply(categorize_income)

**Задание 17. Выведите на экран перечень уникальных целей взятия кредита из столбца `purpose`.**

In [24]:
data['purpose'].unique()

array(['покупка жилья', 'приобретение автомобиля',
       'дополнительное образование', 'сыграть свадьбу',
       'операции с жильем', 'образование', 'на проведение свадьбы',
       'покупка жилья для семьи', 'покупка недвижимости',
       'покупка коммерческой недвижимости', 'покупка жилой недвижимости',
       'строительство собственной недвижимости', 'недвижимость',
       'строительство недвижимости', 'на покупку подержанного автомобиля',
       'на покупку своего автомобиля',
       'операции с коммерческой недвижимостью',
       'строительство жилой недвижимости', 'жилье',
       'операции со своей недвижимостью', 'автомобили',
       'заняться образованием', 'сделка с подержанным автомобилем',
       'получение образования', 'автомобиль', 'свадьба',
       'получение дополнительного образования', 'покупка своего жилья',
       'операции с недвижимостью', 'получение высшего образования',
       'свой автомобиль', 'сделка с автомобилем',
       'профильное образование', 'высшее об

**Задание 18. Создайте функцию, которая на основании данных из столбца `purpose` сформирует новый столбец `purpose_category`, в который войдут следующие категории:**

- `'операции с автомобилем'`,
- `'операции с недвижимостью'`,
- `'проведение свадьбы'`,
- `'получение образования'`.

**Например, если в столбце `purpose` находится подстрока `'на покупку автомобиля'`, то в столбце `purpose_category` должна появиться строка `'операции с автомобилем'`.**

**Используйте собственную функцию с именем `categorize_purpose()` и метод `apply()`. Изучите данные в столбце `purpose` и определите, какие подстроки помогут вам правильно определить категорию.**

In [25]:
def categorize_purpose(row):
    try:
        if 'автом' in row:
            return 'операции с автомобилем'
        elif 'жил' in row or 'недвиж' in row:
            return 'операции с недвижимостью'
        elif 'свад' in row:
            return 'проведение свадьбы'
        elif 'образов' in row:
            return 'получение образования'
    except:
        return 'нет категории'

In [26]:
data['purpose_category'] = data['purpose'].apply(categorize_purpose)

### Шаг 3. Исследуйте данные и ответьте на вопросы

#### 3.1 Есть ли зависимость между количеством детей и возвратом кредита в срок?

In [27]:
data_grouped = data.groupby('children').agg({'debt':['count', 'sum']})  # группируем датафрейм по столбцу "children" и рассчитываем количество должников по столбцу "debt" через 'sum' (значение 1 в 'debt') и общее кол-во записей через 'count'
data_grouped['debt_proportion'] = data_grouped['debt']['sum']/data_grouped['debt']['count']*100  # обращаемся к новому датафрейму и добавляем новый столбец - доля должников по отношению к общему количеству заемщиков
print(data_grouped)

           debt       debt_proportion
          count   sum                
children                             
0         14091  1063        7.543822
1          4808   444        9.234609
2          2052   194        9.454191
3           330    27        8.181818
4            41     4        9.756098
5             9     0        0.000000


**Вывод:** Бездетные заемщики чуть реже выступают должниками по кредитам - в среднем 7.5%. В целом уровень задолженности сохраняется примерно на одном уровне - наблюдается разброс в 2.2 процентных пункта между бездетными заемщиками и земщиками с 1-3 детьми. Среди заемщиков с 5 детьми задолженностей не наблюдается, что вероятно обусловлено прежде всего незначительностью выборки (9 случаев).

#### 3.2 Есть ли зависимость между семейным положением и возвратом кредита в срок?

In [28]:
data_grouped_2 = data.groupby('family_status').agg({'debt':['count', 'sum']})  #  группируем датафрейм по семейному статусу,рассчитываем количество должников по столбцу "debt" через 'sum' и общее кол-во записей через 'count' 
data_grouped_2['debt_proportions'] = data_grouped_2['debt']['sum']/data_grouped_2['debt']['count']*100  # добавляем новый столбец - доля должников по отношению к общему количеству заемщиков
print(data_grouped_2.sort_values(by = 'debt_proportions', ascending = False))

                        debt      debt_proportions
                       count  sum                 
family_status                                     
Не женат / не замужем   2796  273         9.763948
гражданский брак        4134  385         9.313014
женат / замужем        12261  927         7.560558
в разводе               1189   84         7.064760
вдовец / вдова           951   63         6.624606


**Вывод:** Заемщики, не состоящие в официально зарегистрированных отношениях несколько чаще становятся должниками по кредитам: 9.7-9.3% против 7.5% у женатых и замужних заемщиков. Реже всего должниками становятся вдовые заемщики - 6.6%

#### 3.3 Есть ли зависимость между уровнем дохода и возвратом кредита в срок?

In [29]:
data_grouped_3 = data.groupby('total_income_category').agg({'debt':['count', 'sum']})  #  группируем датафрейм по уровню дохода, подсчитываем количество должников (столбец "sum") и общее количество заемщиков ('count')
data_grouped_3['debt_proportions'] = data_grouped_3['debt']['sum']/data_grouped_3['debt']['count']*100  # добавляем расчет доли должников
print(data_grouped_3.sort_values(by = 'debt_proportions', ascending = False))

                            debt       debt_proportions
                           count   sum                 
total_income_category                                  
E: меньше 30 000              22     2         9.090909
C:от 50 000 до 200 000     15921  1353         8.498210
A: больше 1 000 000           25     2         8.000000
B от 200 000 до 1 000 000   5014   354         7.060231
D: от 30 000 до 50 000       349    21         6.017192


**Вывод:** Заемщики со уровнем дохода от 30 до 50 тысяч рублей реже всего выступают должниками по кредитам - всего лишь в 6% случаев. На втором месте должники с уровнем дохода от 200 тысяч рублей до 1 миллиона: 7% случаев. Чаще всего просрочки по кредитам наблюдаются у заемщиков с минимальным уровнем дохода - до 30 тысяч рублей, однако эти данные нельзя считать репрезентативными в силу малочисленности выборки (всего 22 кейса).

#### 3.4 Как разные цели кредита влияют на его возврат в срок?

In [30]:
data_grouped_4 = data.groupby('purpose_category').agg({'debt':['count', 'sum']})  #  группируем датафрейм по цели кредита, подсчитываем количество должников (столбец "sum") и общее количество заемщиков ('count')
data_grouped_4['debt_proportions'] = data_grouped_4['debt']['sum']/data_grouped_4['debt']['count']*100  # добавляем расчет доли должников
print(data_grouped_4.sort_values(by = 'debt_proportions', ascending = False))


                           debt      debt_proportions
                          count  sum                 
purpose_category                                     
операции с автомобилем     4279  400         9.347978
получение образования      3988  369         9.252758
проведение свадьбы         2313  183         7.911803
операции с недвижимостью  10751  780         7.255139


**Вывод:** Кредиты выданные для операций с автомобилями и получение образования чаще выступают просроченными - в среднем в 9.2-9.3 процентов случаев. Кредиты, выданные для операций с недвижимостью не выплачивают реже всего - в 7.2% случаев. 

#### 3.5 Приведите возможные причины появления пропусков в исходных данных.

1. Гипотеза 1: данные не были предоставлены клиентом - не подходит, т.к. в таком случае не был бы выдан кредит. ❌
2. Гипотеза 2: клиент не работает и, соответственно, у него нет дохода и дней трудоустройства  - доп анализ строк с пустыми значениями гипотезу не подтвердил - в графе источник дохода работа указана (сотрудник, пенсионер и другие роли) ❌
3. Гипотеза 3: данные просто не подгрузились по технической или иной причине. 
    В пользу гипотезы 3 говорит тот факт, что количество пропусков идентичное и в столбце "total_income" и в столбце "days_employed". Более того, они принадлежат одним и тем же строкам. 
    
    Количество пропусков - 2174 (из 21525 записей, примерно 10%) , значимых корреляций с видом деятельности, целью кредита, семейным положением мною не выявлено. Возможно имело место региональная или иная специфика - недостаточно данных для такого анализа.

#### 3.6 Объясните, почему заполнить пропуски медианным значением — лучшее решение для количественных переменных.

*Ответ:* При необходимости заполнить пропущенные данные представляется более эффективным использовать медианные, а не средние значения. В таком случае мы застрахованы от влияния выбросов (резко отличающихся данных) на общую выборку. Например при заполнении пропусков в данных, связанных со уровнем дохода, стоимостью продуктов и т.п. - достаточно одного кейса со значительным отклонением от других значений (например зарплата топ-менеджера среди общих зп сотрудников большого предприятия) для сильного искажения средних показателей. Использование медианы - срединного значения (ровно половина данных ниже его, а другая- выше) в этом случае более безопасно с точки зрения заполнения пропусков.

### Шаг 4: общий вывод.

Анализ данных не выявил критичных корреляций между задержкой с выплатой кредита и семейным статусом/наличием детей, целью дохода или уровнем дохода заемщика. Между тем отмечено несколько тенденций:
1. Бездетные заемщики реже выступают должниками, чем заемщики с детьми. Одновременно с этим просрочка кредита чаще наблюдается у заемщиков, не состоящих в зарегистированных отношениях (семейный статус - холост или гражданский брак), что выглядит несколько противоречиво. 
Дополнительный кросс-анализ наличия задолженности (расчеты ниже) у заемщика в разрезе количества детей и семейного статуса выявил более явные закономерности: самый высокий уровень задолженности (свыше 10%) отмечен у заемщиков с детьми и и при этом не состоящих в зарегистированных отношениях(холост, в разводе, вдовец), т.е потенциально выступающими единственными родителями. 
Можно выдвинуть гипотезу о более высоком риске просроченных кредитов у заемщиков - соло родителей, возможно связанный с финансовым грузом содержания ребенка.
2. Анализ соотношения уровня дохода и задолженности по кредиту показал что реже всего долги по кредитам у заемщиков с уровнем дохода от 30 до 50 тысяч рублей - всего  6%. 
На втором месте должники с уровнем дохода от 200 тысяч рублей до 1 миллиона: 7% случаев. Чаще всего просрочки по кредитам наблюдаются у заемщиков с минимальным уровнем дохода - до 30 тысяч рублей, однако эти данные нельзя считать репрезентативными в силу малочисленности выборки (всего 22 кейса). На втором месте по просрочкам заемщики с уровнем дохода от 50 до 200 тысяч рублей - 8.5%
3. Чаще всего просроченными являются кредиты выданные для операций с автомобилями и получением образования - в среднем в 9.2-9.3 процентов случаев. Кредиты, выданные для операций с недвижимостью не выплачивают реже всего - в 7.2% случаев.
